macro-circom/src/code_gen/check_utxo_code.rs

```diff
@@ -0,0 +1,940 @@
+            checkInstructionHash{{UtxoName}}[inUtxoIndex].enabled <== is{{is_In}}AppUtxo{{UtxoName}}[inUtxoIndex];
+        }
+        {{/if}}
+        component {{utxoName}} = {{this.utxoType}}();
+
+        {{#with this}}{{#each utxoData}}
+        {{../../utxoName}}.{{this.input}}In <== {{../../utxoName}}{{this.Input}};
+        {{/each}}{{/with}}
+        {{#with this}}{{#each nativeUtxoFields}}
+        signal input {{../../utxoName}}{{this.InputField}};
+        {{../../utxoName}}.{{this.inputField}}In <== {{../../utxoName}}{{this.InputField}};
+        {{/each}}{{/with}}
+        {{../../utxoName}}.utxoDataHashIn <== {{isAppUtxo}};
+        
+        component {{../../utxoName}}AmountHasher = Poseidon(2);
+        {{../../utxoName}}AmountHasher.inputs[0] <== 0;
+        {{../../utxoName}}AmountHasher.inputs[1] <== {{../../utxoName}}.assetSpl;
+
+        component {{../../utxoName}}AssetHasher = Poseidon(2);
+        {{../../utxoName}}AssetHasher.inputs[0] <== {{../../utxoName}}.amountSol;
+        {{../../utxoName}}AssetHasher.inputs[1] <== {{../../utxoName}}.amountSpl;
+
+        component {{../../utxoName}}UtxoCheckHasher = Poseidon(8);
+        {{../../utxoName}}UtxoCheckHasher.inputs[0] <== 0; // TxVersion
+        {{../../utxoName}}UtxoCheckHasher.inputs[1] <== {{../../utxoName}}AmountHasher.out;
+        {{../../utxoName}}UtxoCheckHasher.inputs[2] <== {{../../utxoName}}AssetHasher.out;
+        {{../../utxoName}}UtxoCheckHasher.inputs[3] <== {{../../utxoName}}.blinding;
+        {{../../utxoName}}UtxoCheckHasher.inputs[4] <== {{../../utxoName}}AssetHasher.out;
+        {{../../utxoName}}UtxoCheckHasher.inputs[5] <== {{../../utxoName}}.utxoDataHash;
+        {{../../utxoName}}UtxoCheckHasher.inputs[6] <== 0;
+        {{../../utxoName}}UtxoCheckHasher.inputs[7] <== {{../../utxoName}}.pspOwner;
+
+
+"#;
+        // TODO: make direct part of transaction hash
+        // - put check at the end of the transaction
+        // - don't compute utxo hashes of non-declared utxos just pass in the commitments
+        let mut all_utxo_data = Vec::<handlebars::JsonValue>::new();
+        for utxo_data_field in &self.type_struct.as_ref().unwrap().fields {
+            all_utxo_data.push(serde_json::json!({
+                "component": format!("UtxoData{}", utxo_data_field.clone().to_upper_camel_case()),
+                "input": utxo_data_field,
+                "Input": utxo_data_field.to_upper_camel_case(),
+            }));
+        }
+
+        // This handles the case that we want to check a utxo that is not a program utxo the utxo data is zero
+        let len_utxo_data = if self.type_struct.as_ref().is_some() {
+            self.type_struct.as_ref().unwrap().fields.len()
+        } else {
+            0
+        };
+
+        let native_utxo_fields = vec![
+            "publicKey",
+            "blinding",
+            "pspOwner",
+            // "utxoDataHash",
```

Comment:
I see it uncommented in one other file. Is there any reason why we don't handle `utxoDataHash` here?

Reply:
It's handled separately in line 242, because for a native utxo we enforce it as zero and for app utxos we use utxoDataHasher.out.